builtin/logical/mongodb/path_roles.go

```diff
@@ -0,0 +1,182 @@
+			},
+			"roles": &framework.FieldSchema{
+				Type:        framework.TypeString,
+				Description: "MongoDB roles to assign to the users generated for this role.",
+			},
+		},
+
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.ReadOperation:   b.pathRoleRead,
+			logical.UpdateOperation: b.pathRoleCreate,
+			logical.DeleteOperation: b.pathRoleDelete,
+		},
+
+		HelpSynopsis:    pathRoleHelpSyn,
+		HelpDescription: pathRoleHelpDesc,
+	}
+}
+
+func (b *backend) Role(s logical.Storage, n string) (*roleEntry, error) {
+	entry, err := s.Get("role/" + n)
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return nil, nil
+	}
+
+	var result roleEntry
+	if err := entry.DecodeJSON(&result); err != nil {
+		return nil, err
+	}
+
+	return &result, nil
+}
+
+func (b *backend) pathRoleDelete(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	err := req.Storage.Delete("role/" + data.Get("name").(string))
+	if err != nil {
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+func (b *backend) pathRoleRead(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	role, err := b.Role(req.Storage, data.Get("name").(string))
+	if err != nil {
+		return nil, err
+	}
+	if role == nil {
+		return nil, nil
+	}
+
+	rolesJsonBytes, err := json.Marshal(role.MongoDBRoles)
+	if err != nil {
+		return nil, err
+	}
+
+	return &logical.Response{
+		Data: map[string]interface{}{
+			"db": role.DB,
+			"roles": string(rolesJsonBytes),
+		},
+	}, nil
+}
+
+func (b *backend) pathRoleList(
+	req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	entries, err := req.Storage.List("role/")
+	if err != nil {
+		return nil, err
+	}
+
+	return logical.ListResponse(entries), nil
+}
+
+func (b *backend) pathRoleCreate(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	name := data.Get("name").(string)
+	roleDB := data.Get("db").(string)
+
+	// Example roles JSON: [ "readWrite", { "role": "readWrite", "db": "test" } ]
+	var roles []interface{}
+	rolesJson := []byte(data.Get("roles").(string))
+	if len(rolesJson) > 0 {
+		var rolesArray []interface{}
+		err := json.Unmarshal(rolesJson, &rolesArray)
+		if err != nil {
+			return nil, err
+		}
+		for _, rawItem := range rolesArray {
+			switch item := rawItem.(type) {
+			case string:
+				roles = append(roles, item)
+			case map[string]interface{}:
+				if db, ok := item["db"].(string); ok {
+					if r, ok := item["role"].(string); ok {
+						roles = append(roles, mongodbRole{Role: r, DB: db})
+					}
+				}
+			}
+		}
+	}
+
+	// Store it
+	entry, err := logical.StorageEntryJSON("role/"+name, &roleEntry{
+		DB:    roleDB,
+		MongoDBRoles: roles,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if err := req.Storage.Put(entry); err != nil {
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+type roleEntry struct {
+	DB           string        `json:"db"`
+	MongoDBRoles []interface{} `json:"roles"`
```

Comment:
I must admit, I don't really understand what you're storing here.

From https://docs.mongodb.com/manual/tutorial/manage-users-and-roles/ it appears to me that when granting permissions, you first have to have a user; then, you have an array of dictionaries that you grant to that user, where the two keys are the db and the role. However, we don't know the user at this point, so we are simply trying to define the DB/role mappings.

As a result I don't really understand why you'd ever have something that is a pure string in the data (as shown in your comment) -- as far as I can tell the roles parameter ought to be able to be unmarshaled from a string directly into a slice of `mongodbRole`, so in `roleEntry` you'd have `MongoDBRoles []mongodbRole`. Am I missing something here?